utils.go

```diff
@@ -14,17 +14,17 @@ func contains(items []string, value string) bool {
 	return false
 }
 
-func convert(i interface{}) interface{} {
+func unmarshalledYAMLToMarshallableJSON(i interface{}) interface{} {
 	switch x := i.(type) {
 	case map[interface{}]interface{}:
 		m2 := map[string]interface{}{}
 		for k, v := range x {
-			m2[k.(string)] = convert(v)
+			m2[k.(string)] = unmarshalledYAMLToMarshallableJSON(v)
 		}
 		return m2
 	case []interface{}:
 		for i, v := range x {
-			x[i] = convert(v)
+			x[i] = unmarshalledYAMLToMarshallableJSON(v)
 		}
 	}
 	return i
```

Comment:
Cool but kinda makes my head hurt. Can you elaborate on why this is needed? There is a yaml to json function and an unmarshal in `github.com/ghodss/yaml`.

Reply:
Awesome finding! I'll try to use that one instead

Reply:
The `github.com/ghodss/yaml` doesn't support non-string map keys, so this function is useless. The type will never be `map[interface{}]interface{}`.

The YAML parser used in wharf-core, `gopkg.in/yaml.v2`, does support non-string map keys.

Example:

```go
package main

import (
	"fmt"

	wharfAPIYAML "github.com/ghodss/yaml"
	wharfCoreYAML "gopkg.in/yaml.v2"
)

func main() {
	var yamlBytes = []byte(`123: foo`)

	var obj1 interface{}
	wharfAPIYAML.Unmarshal(yamlBytes, &obj1)
	fmt.Printf("%#v\n", obj1)

	var obj2 interface{}
	wharfCoreYAML.Unmarshal(yamlBytes, &obj2)
	fmt.Printf("%#v\n", obj2)
}

// Output:
// map[string]interface {}{"123":"foo"}
// map[interface {}]interface {}{123:"foo"}
```

This function is older than this repo, so my only guess is that there was perhaps a dependency switch, but the function was not removed.

---

Q: Do we either:

1. Remove `unmarshalledYAMLToMarshallableJSON` as `github.com/ghodss/yaml` doesn't need it?

   The use case here is just to parse the build configuration so the frontend can use a pre-parsed JSON object instead of having to do any YAML parsing itself. As JSON doesn't support non-string keys, this "drawback" from `github.com/ghodss/yaml` could actually be seen as a strength.

2. Refactor `unmarshalledYAMLToMarshallableJSON` to be easier to understand and switch to `gopkg.in/yaml.v2`?

   Better to use the same YAML serializer throughout, instead of having multiple in the same project.

Reply:
I would prefer 1.

Reply:
I've gone for option 1 and added some tests to verify that it works. I also changed the MarshalJSON to be a non-pointer method as that did otherwise only work for pointer values.